### DISCUSSION TITLE
Best way in CSLA to enforce one of two boolean properties are true?

### DISCUSSION THREAD

**Discussion by MarkOverstreet** [MarimerLLC/csla#1744] [Questions]
Pretty new to CSLA so I'm sure this will be easy for someone but I have a simple root object defined and I have two properties as seen below:

        public static readonly PropertyInfo<bool> DoesExpireProperty = RegisterProperty<bool>(nameof(DoesExpire));
        [Display(Name = "Does This Expire?")]
        public bool DoesExpire
        {
            get =>  GetProperty(DoesExpireProperty); 
            set =>  SetProperty(DoesExpireProperty, value);
        }

        public static readonly PropertyInfo<bool> DoesAccumulateHoursProperty = RegisterProperty<bool>(nameof(DoesAccumulateHours));
        [Display(Name = "Does This Accumulate Hours?")]
        public bool DoesAccumulateHours
        {
            get => GetProperty(DoesAccumulateHoursProperty);
            set => SetProperty(DoesAccumulateHoursProperty, value);
        }

I simply want to make sure that one and only one of these are true at any given time but I don't really want to create a business rule and let the user select both only to get an error.  I'd prefer to just set one to false automatically if the other is set to true and vice versa.   I tried to add some code to the Setter to do this but that causes some strange behavior along with some analyzer messages complaining to me not do to that in CSLA.   

So what would be the best way to implement that?

Thanks
Mark

**Comment by rockfordlhotka** [comment]
The short version (b/c I'm not at a computer) is that you create a rule that has a primary and affected property as inputs to the constructor.

In the rule you get the primary property value, and based on your rule, you might change the affected property using `context.AddOutValue`.

Then add the rule to each of your properties, with the other property as the affected property. 

The one thing to watch for is that you want to make sure your logic doesn't create an infinite loop, where changing one property causes a change to the other, and the change to the other _also_ causes a change to the first one - clearly a bad thing!! 😊

**Reply by rockfordlhotka** [reply]
Here's some code to consider.

```c#
using System;
using Csla;
using Csla.Rules;

namespace ConsoleApp4
{
  class Program
  {
    static void Main(string[] args)
    {
      var obj = DataPortal.Create<RulesTest>();
      Console.WriteLine($"Flag 1 is {obj.Flag1}; Flag 2 is {obj.Flag2}");
      obj.Flag1 = true;
      Console.WriteLine($"Flag 1 is {obj.Flag1}; Flag 2 is {obj.Flag2}");
      obj.Flag1 = false;
      Console.WriteLine($"Flag 1 is {obj.Flag1}; Flag 2 is {obj.Flag2}");
      obj.Flag2 = false;
      Console.WriteLine($"Flag 1 is {obj.Flag1}; Flag 2 is {obj.Flag2}");
    }
  }

  [Serializable]
  public class RulesTest : BusinessBase<RulesTest>
  {
    public static readonly PropertyInfo<bool> Flag1Property = RegisterProperty<bool>(nameof(Flag1));
    public bool Flag1
    {
      get => GetProperty(Flag1Property);
      set => SetProperty(Flag1Property, value);
    }

    public static readonly PropertyInfo<bool> Flag2Property = RegisterProperty<bool>(nameof(Flag2));
    public bool Flag2
    {
      get => GetProperty(Flag2Property);
      set => SetProperty(Flag2Property, value);
    }

    protected override void AddBusinessRules()
    {
      base.AddBusinessRules();
      BusinessRules.AddRule(new SyncFlags(Flag1Property, Flag2Property));
      BusinessRules.AddRule(new SyncFlags(Flag2Property, Flag1Property));
    }

    [Create]
    [RunLocal]
    private void Create()
    {
      BusinessRules.CheckRules();
    }
  }

  public class SyncFlags : Csla.Rules.BusinessRule
  {
    private Csla.Core.IPropertyInfo _target;
    public SyncFlags(Csla.Core.IPropertyInfo primary, Csla.Core.IPropertyInfo target)
      :base(primary)
    {
      _target = target;
      AffectedProperties.Add(target);
      InputProperties.Add(primary);
      InputProperties.Add(target);
    }

    protected override void Execute(IRuleContext context)
    {
      var p = (bool)context.InputPropertyValues[PrimaryProperty];
      var t = (bool)context.InputPropertyValues[_target];
      if (p && t)
        context.AddOutValue(_target, !t);
      else if (!p && !t)
        context.AddOutValue(_target, !t);
    }
  }
}
```

**Reply by MarkOverstreet** [reply]
Thanks for the advice and code.  I'll try this out!

**Reply by MarkOverstreet** [reply]
Ok that works as far as enforcing the two boolean properties cannot be true at the same time.  But I've run into a screen refresh issue.  I am reading your Blazor book and maybe I've found a bug or maybe I just don't know what I'm doing :)  

I simply have a Category class with this definition as...

`_public class Category : BusinessBase<Category>_`

and a collection of those as a DynamicallyListBase as ...

`_public class Categories : DynamicListBase<Category>_`


because from what i understand that must be the type if I am binding root objects to a grid for editing.  

I also used your viewmodel as ...

`@inject ViewModel<Categories> vm;`

and this...

    protected override void OnInitialized()
    {
        vm.ModelPropertyChanged += async (s, e) => await InvokeAsync(() => StateHasChanged());


However, when the business rule you provided changes one of the boolean values, this code never fires to force a screen update.

Any thoughts as to why the ModelPropertyChanged event is never called?  Is it because I am using the DynamicListBase?  

Thanks for the assistance.

**Reply by IgnyteSoftware** [reply]
In fact, it appears that the ModelPropertyChanged event never fires for any business object property changing?  Is this a known issue?
